File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/expression/Value.scala
package io.github.chaosunity.casc.parsing.expression

import io.github.chaosunity.casc.parsing.`type`.Type

class Value(_type: Type,
            val value: String,
            _negative: Boolean) extends Expression {
    override def `type`: Type = _type

    override def negative: Boolean = _negative
}

File: src/main/kotlin/org/casclang/casc/parsing/node/expression/Value.kt
package org.casclang.casc.parsing.node.expression

import org.casclang.casc.parsing.type.BuiltInType

data class Value(override val type: BuiltInType, val value: String) : FoldableExpression<Value>(type) {
    fun evaluate(targetType: BuiltInType = type): Number =
        when (targetType) {
            BuiltInType.DOUBLE -> value.toDouble()
            BuiltInType.FLOAT -> value.toFloat()
            BuiltInType.LONG -> value.toLong()
            BuiltInType.INT -> value.toInt()
            else -> throw IllegalArgumentException()
        }
}
